bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -206,177 +262,232 @@ private void closeConnection() throws RemoteControllerException {
-            } catch (RemoteControllerException e) {
-                logger.debug("Couldn't send command", e);
-                logger.debug("Retry one time...");
-
-                closeConnection();
-                openConnection();
-
-                sendKeyData(key, false);
-            }
+    public void sendSourceApp(final String appName, @Nullable String url) {
+        Stream<Map.Entry<String, App>> st = (noApps()) ? manApps.entrySet().stream() : apps.entrySet().stream();
+        boolean found = st.filter(a -> a.getKey().equals(appName) || a.getValue().name.equals(appName))
+                .map(a -> sendSourceApp(a.getValue().appId, a.getValue().type == 2, url)).findFirst().orElse(false);
+        if (!found) {
+            // treat appName as appId with optional type number eg "3201907018807, 2"
+            String[] appArray = (url == null) ? appName.trim().split(",") : "org.tizen.browser,4".split(",");
+            sendSourceApp(appArray[0].trim(), (appArray.length > 1) ? "2".equals(appArray[1].trim()) : true, url);
+        }
+    }
 
-            if ((keys.size() - 1) != i) {
-                // Sleep a while between commands
-                try {
-                    Thread.sleep(sleepInMs);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    return;
-                }
+    public boolean sendSourceApp(String appId, boolean type, @Nullable String url) {
+        if (noApps()) {
+            // 2020 TV's and later use webSocketV2 for app launch
+            webSocketV2.sendSourceApp(appId, type, url);
+        } else {
+            if (webSocketV2.isConnected() && url == null) {
+                // it seems all Tizen TV's can use webSocketV2 if it connects
+                webSocketV2.sendSourceApp(appId, type, url);
+            } else {
+                webSocketRemote.sendSourceApp(appId, type, url);
             }
         }
-
-        logger.debug("Command(s) successfully sent");
+        return true;
     }
 
-    private void sendKeyData(KeyCode key, boolean press) throws RemoteControllerException {
-        webSocketRemote.sendKeyData(press, key.toString());
+    public void sendUrl(String url) {
+        String processedUrl = url.replace("/", "\\/");
+        sendSourceApp("Internet", processedUrl);
     }
 
-    public void sendSourceApp(String app) {
-        String appName = app;
-        App appVal = apps.get(app);
-        boolean deepLink = false;
-        appName = appVal.appId;
-        deepLink = appVal.type == 2;
-
-        webSocketRemote.sendSourceApp(appName, deepLink);
+    public boolean closeApp() {
+        return webSocketV2.closeApp();
     }
 
-    public void sendUrl(String url) {
-        String processedUrl = url.replace("/", "\\/");
-        webSocketRemote.sendSourceApp("org.tizen.browser", false, processedUrl);
+    /**
+     * Get app status after 3 second delay (apps take 3s to launch)
```

Comment:
It shouldn't matter, apps all take a small amount of time to load (2 ish seconds on my TV). If the app takes longer to load, it will be picked up on the next poll (maximum of 10 seconds).